### PR TITLE
Issue 209

### LINKED ISSUE #209: relators having same domain and range loses range
After publishing WS, deleting it, creating new one and importing the same data, the relators with the same domain and range (e.g. má první verzi dokumentu) lose range (má vztažený prvek 2). If I create a new one, it has predefined default cardinality (IDKY).

### PR DESCRIPTION
**Pull request by bindeali**
## Info:
* Closes #209.
* Can be merged right away.
## Deploy previews:
[ML Test](https://deploy-preview-223--ontographer.netlify.app/?workspace=https://slovn%C3%ADk.gov.cz/datov%C3%BD/pracovn%C3%AD-prostor/pojem/metadatov%C3%BD-kontext/instance-1101091271)
[Datové fondy VS](https://deploy-preview-223--ontographer.netlify.app/?workspace=https://slovn%C3%ADk.gov.cz/datov%C3%BD/pracovn%C3%AD-prostor/pojem/metadatov%C3%BD-kontext/instance-1153182747)
[Marie test](https://deploy-preview-223--ontographer.netlify.app/?workspace=https://slovn%C3%ADk.gov.cz/datov%C3%BD/pracovn%C3%AD-prostor/pojem/metadatov%C3%BD-kontext/instance-530560086)
[mm-test](https://deploy-preview-223--ontographer.netlify.app/?workspace=https://slovn%C3%ADk.gov.cz/datov%C3%BD/pracovn%C3%AD-prostor/pojem/metadatov%C3%BD-kontext/instance-2000996011)
[Obchodní rejstřík](https://deploy-preview-223--ontographer.netlify.app/?workspace=https://slovn%C3%ADk.gov.cz/datov%C3%BD/pracovn%C3%AD-prostor/pojem/metadatov%C3%BD-kontext/instance201242184)
[Sčítání lidu](https://deploy-preview-223--ontographer.netlify.app/?workspace=https://slovn%C3%ADk.gov.cz/datov%C3%BD/pracovn%C3%AD-prostor/pojem/metadatov%C3%BD-kontext/instance-2011762727)
[OFN číselníky a klasifikace](https://deploy-preview-223--ontographer.netlify.app/?workspace=https://slovn%C3%ADk.gov.cz/datov%C3%BD/pracovn%C3%AD-prostor/pojem/metadatov%C3%BD-kontext/instance-1810443144)
[Sbírka](https://deploy-preview-223--ontographer.netlify.app/?workspace=https://slovn%C3%ADk.gov.cz/datov%C3%BD/pracovn%C3%AD-prostor/pojem/metadatov%C3%BD-kontext/instance1483547911)
[časoprostor](https://deploy-preview-223--ontographer.netlify.app/?workspace=https://slovn%C3%ADk.gov.cz/datov%C3%BD/pracovn%C3%AD-prostor/pojem/metadatov%C3%BD-kontext/instance926567775)

### CODE REVIEW

**Review by psiotwo — COMMENTED** [review]
lgtm, tested on new vocabulary with eSbírka. Yet not sure whether I replicate the scenario correctly. @MichalMed is on vacation this week, so we might wait for his approval here.

File: src/function/FunctionRestriction.ts
Comment (karelklima) [inline src/function/FunctionRestriction.ts:7]: Two points regarding `obj: { [key: string]: any }`:

1) There is a handy Record utility type that can be used in similar cases
https://www.typescriptlang.org/docs/handbook/utility-types.html#recordkeystype

2) The `any` would deserve better specification, I am sure the objects types that can come in are known so it would be great to have it replaced i.e. by "SomeType | OtherType | MaybeAThirdType" insted. Alternatively it's possible to describe the interface, e.g. `Record<string, { restrictions: Restriction[] }>`

File: src/function/FunctionRestriction.ts
Comment (karelklima) [inline src/function/FunctionRestriction.ts:13]: It would be probably worth building a dedicated comparator function here - something like `Restriction.isSame(other: Restriction)`.

File: src/function/FunctionRestriction.ts
Comment (karelklima) [inline src/function/FunctionRestriction.ts:33]: There is this pattern that is copied on multiple places:
`Object.keys(ProjectElements).find(element => ProjectElements[element].iri === XXXXX);`

It could be worth refactor it to a separate utility method, e.g.
`const findElementByIri = (iri: string) => Object.keys(ProjectElements).find(element => ProjectElements[element].iri === iri)`
and then just do something like
`const domainID = findElementByIri(iri)`
It would improve readability a lot!

File: src/config/RestrictionConfig.ts
Comment (bindetad) [reply src/config/RestrictionConfig.ts]: Yes, that is quite an unnecessary oversight. I changed it and will keep an eye on it in future PRs. Thank you for the comment.

File: src/function/FunctionRestriction.ts
Comment (bindetad) [reply src/function/FunctionRestriction.ts]: Implemented. Thank you for the comment.

File: src/function/FunctionRestriction.ts
Comment (bindetad) [reply src/function/FunctionRestriction.ts]: Implemented. Thank you for the comment.

File: src/function/FunctionRestriction.ts
Comment (bindetad) [reply src/function/FunctionRestriction.ts]: I eventually replaced it with just Restriction[], but I will keep Records in mind - there are other places that will need it. Thank you for the comment.

**Review by karelklima — APPROVED** [review]
Looks much better, thanks for all the changes!

**Comment by psiotwo** [comment]
> Looks much better, thanks for all the changes!

Great job guys, thanks both !

**Comment by MichalMed** [comment]
for replication, I just need to access the mission control, create new project, create some relations, than delete the project, create new one and upload relations created in previously deleted project. I do not know how to do it here.

**Comment by bindetad** [comment]
> for replication, I just need to access the mission control, create new project, create some relations, than delete the project, create new one and upload relations created in previously deleted project. I do not know how to do it here.

You can create the new workspace as normal, then change the base URL from "Editovat vztahy" from 
`ontographer.netlify.app/?workspace=[workspaceIRI]` 
to 
`deploy-preview-223--ontographer.netlify.app/?workspace=[workspaceIRI]`

**Review by MichalMed — COMMENTED** [review]
lgtm

**Pull request merged**
